### ISSUE TITLE
cannot reproduce results on Coinrun.

### ISSUE THREAD

**Issue by sjtuytc** [openai/train-procgen#6]
I run the following command:
`python -m train_procgen.train --env_name coinrun --num_levels 500`
But I cannot reproduce the results as you've presented. Actually the final training score is around 5, but you report as over 7.
I attach my progress.csv in this issue. Can you help me out?
[coinrun_progress.pdf](https://github.com/openai/train-procgen/files/4065891/coinrun_progress.pdf)

**Comment by kcobbe** [comment]
Compared to our baseline, your batch size is 8x too small. Specifically, we average gradients across 8 workers using MPI, where each worker is using 64 copies of the environment (just like your single worker). You can try `mpiexec -np 8 python -m train_procgen.train --env_name coinrun --num_levels 500`.

Edit: I should have said 4 workers instead of 8, so the command should be `mpiexec -np 4 python -m train_procgen.train --env_name coinrun --num_levels 500`. Thanks @GNiendorf.

**Comment by GNiendorf** [comment]
Small thing, but in the paper you say that you use 4 workers for PPO (and 8 for rainbow, not in this code). Is that an error? @kcobbe

**Comment by Mckiev** [comment]
I’ve trained an agent on coinrun for 200M timesteps, specifying num_envs 512, nminibatches 64 (to be able to have same effective batch size without mpi). My agent quickly achieves 6-7 average reward and never improved further. However  on several other environments the training curves are about same as reported in the paper

**Comment by dipamc** [comment]
> I’ve trained an agent on coinrun for 200M timesteps, specifying num_envs 512, nminibatches 64 (to be able to have same effective batch size without mpi). My agent quickly achieves 6-7 average reward and never improved further. However on several other environments the training curves are about same as reported in the paper

Scaling up the nminibatches along with num_envs effectively keeps the same batch size, whereas using mpi effectively increases the batch size. If you want to have the same effect as using mpi you have to get gradients for 8 minibatches (with your settings at 512 envs and 64 nminibatches), average them and then apply to the optimizer.

**Comment by sjtuytc** [comment]
Let me explain what you said. For every batch there is always 8 data for each env, if you only scale the minibatches up. But it would be better you have one env with more data while averaging gradients by utilizing those data.
However, it costs significantly more CPUs.
> > I’ve trained an agent on coinrun for 200M timesteps, specifying num_envs 512, nminibatches 64 (to be able to have same effective batch size without mpi). My agent quickly achieves 6-7 average reward and never improved further. However on several other environments the training curves are about same as reported in the paper
> 
> Scaling up the nminibatches along with num_envs effectively keeps the same batch size, whereas using mpi effectively increases the batch size. If you want to have the same effect as using mpi you have to get gradients for 8 minibatches (with your settings at 512 envs and 64 nminibatches), average them and then apply to the optimizer.

**Comment by Mckiev** [comment]
> > I’ve trained an agent on coinrun for 200M timesteps, specifying num_envs 512, nminibatches 64 (to be able to have same effective batch size without mpi). My agent quickly achieves 6-7 average reward and never improved further. However on several other environments the training curves are about same as reported in the paper
> 
> Scaling up the nminibatches along with num_envs effectively keeps the same batch size, whereas using mpi effectively increases the batch size. If you want to have the same effect as using mpi you have to get gradients for 8 minibatches (with your settings at 512 envs and 64 nminibatches), average them and then apply to the optimizer.

Thank you for taking time to reply to my comment!
From looking at the code in [ppo2.py](baselines/ppo2/ppo2.py) line 183 is `lossvals = np.mean(mblossvals, axis=0)`, so the gradient is averaged across all `nmininbatches` and applied only once each of the `num_envs` environments made `nsteps`. The only one source of the problem I can see so far is that gradients from each minibatch normalize advantages based on this particular mininbatch. 
I guess I'll have to figure out how to use MPI. So far I struggle to run on single GPU, it is not getting used at all for some reason

**Comment by dipamc** [comment]
> > > I’ve trained an agent on coinrun for 200M timesteps, specifying num_envs 512, nminibatches 64 (to be able to have same effective batch size without mpi). My agent quickly achieves 6-7 average reward and never improved further. However on several other environments the training curves are about same as reported in the paper
> > 
> > 
> > Scaling up the nminibatches along with num_envs effectively keeps the same batch size, whereas using mpi effectively increases the batch size. If you want to have the same effect as using mpi you have to get gradients for 8 minibatches (with your settings at 512 envs and 64 nminibatches), average them and then apply to the optimizer.
> 
> Thank you for taking time to reply to my comment!
> From looking at the code in [ppo2.py](baselines/ppo2/ppo2.py) line 183 is `lossvals = np.mean(mblossvals, axis=0)`, so the gradient is averaged across all `nmininbatches` and applied only once each of the `num_envs` environments made `nsteps`. The only one source of the problem I can see so far is that gradients from each minibatch normalize advantages based on this particular mininbatch.
> I guess I'll have to figure out how to use MPI. So far I struggle to run on single GPU, it is not getting used at all for some reason

No, the lossvals on line 183 is used for final logging only.

I'll clarify. The runner.run on line 142 of ppo2.py generates rollouts of length "nsteps" (256) on all environments (suppose "num_envs" = 512). To simplify the example lets say you have 5 environments and "nsteps" 10. Then the rollouts you have is

env0_step0, env0_step1 ...... env0_step9
env1_step0, env1_step1 ...... env1_step9
....
env4_step0, env4_step1 ...... env4_step9

Then the returns are calculated on the rollouts, so no matter how many environments you have the advantage estimates will be correct. They will only change if you change the rollout length.

Now the data is reordered in this manner (see function sf01 in the runner)
env0_step0, env1_step0, env2_step0 ....... , env2_step9, env3_step9, env4_step9 -> This is passed back to ppo2 as the complete data. 

Then this data is separated into "nminibatches", each minibatch is randomized and used to update the network parameters using model.train (Refer to line 153 to 166 in ppo2.py). For each minibatch, mbloss is found and gradients are applied. This is done for the entire data that came for the runner for "noptepochs" (3 is the default used in train-procgen)

As for normalizing advantages, you can use nminibatches=8, normalize the advantages for the whole minibatch, then further slice the minibatch and compute the gradients of the slice, finally average all the gradients of the slices. (Note this is only possible because the network doesn't have batchnormalization - like with most RL models)

I hope this was helpful, please do point out any mistakes as I'm also a beginner, and others finding out my mistakes is immensely valuable to for me learning better.

**Comment by max-reuter** [comment]
If I want to reproduce results using 8 workers (`mpiexec -np 8`), what would be the appropriate `num_envs`, `timesteps_per_proc`, and `nminibatches`? Or would I need to go beyond these variables and change something in the code?

**Comment by max-reuter** [comment]
@christopherhesse could you please reopen this? ^

**Comment by christopherhesse** [comment]
@max-reuter If `mpiexec -np 4 python -m train_procgen.train --env_name coinrun --num_levels 500` doesn't work for you I can reopen this.  Otherwise please file a new issue.

To change from 4 workers to 8, take `num_envs` and `timesteps_per_proc` and multiply by 4/8, `nminibatches` should likely remain the same.  You may get slightly different results if there is per-process normalization going on.
